File: src/main/java/io/vproxy/vfx/util/algebradata/XYData.java
package io.vproxy.vfx.util.algebradata;

public class XYData implements AlgebraData<XYData> {
    public final double x;
    public final double y;

    public XYData(double x, double y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public XYData plus(XYData other) {
        return new XYData(x + other.x, y + other.y);
    }

    @Override
    public XYData minus(XYData other) {
        return new XYData(x - other.x, y - other.y);
    }

    @Override
    public XYData multiply(double v) {
        return new XYData(x * v, y * v);
    }

    @Override
    public XYData dividedBy(double v) {
        return new XYData(x / v, y / v);
    }

    @Override
    public String toString() {
        return "XYData(" + x + ", " + y + ")";
    }
}

File: src-ts/src/util/algebradata/XYData.ts
import { AlgebraData } from './AlgebraData.js';

export class XYData implements AlgebraData<XYData> {
  constructor(public readonly x: number, public readonly y: number) {}

  plus(other: XYData): XYData { return new XYData(this.x + other.x, this.y + other.y); }
  minus(other: XYData): XYData { return new XYData(this.x - other.x, this.y - other.y); }
  multiply(v: number): XYData { return new XYData(this.x * v, this.y * v); }
  dividedBy(v: number): XYData { return new XYData(this.x / v, this.y / v); }

  toString(): string { return `XYData(${this.x}, ${this.y})`; }
}
